core/common/iterator/src/main/java/org/eclipse/rdf4j/common/iteration/DualUnionIteration.java

```diff
@@ -0,0 +1,89 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Eclipse RDF4J contributors.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Distribution License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ ******************************************************************************/
+
+package org.eclipse.rdf4j.common.iteration;
+
+/**
+ * Provides a bag union of the two provided iterations.
+ */
+public class DualUnionIteration<E, X extends Exception> extends LookAheadIteration<E, X> {
```

Comment:
Union of two iterations is the most common because we usually have an explicit dataset and an inferred one. So it's worth while optimizing for this use case.